ext/surro-gate/selector_ext.c

```diff
@@ -89,69 +75,54 @@ static void SurroGate_Selector_free(int *selector) {
 }
 
 static VALUE SurroGate_Selector_initialize(VALUE self, VALUE logger) {
-  VALUE mConcurrent = rb_const_get(rb_cObject, rb_intern("Concurrent"));
-  VALUE cArray = rb_const_get(mConcurrent, rb_intern("Array"));
-
-  rb_iv_set(self, "@pairing", rb_class_new_instance(0, NULL, cArray)); // @pairing = Concurrent::Array.new
+  rb_iv_set(self, "@scoreboard", rb_class_new_instance(0, NULL, cSurroGate_Scoreboard)); // @scoreboard = Scoreboard.new
   rb_iv_set(self, "@logger", logger);
 
   return Qnil;
 }
 
 static VALUE SurroGate_Selector_push(VALUE self, VALUE left, VALUE right) {
-  int index_ltr, index_rtl, *selector;
-
-  VALUE pair_LTR[2] = {left, right};
-  VALUE pair_RTL[2] = {right, left};
-  VALUE left_to_right = rb_class_new_instance(2, pair_RTL, cSurroGate_Pair); // SurroGate::Pair.new(left, right)
-  VALUE right_to_left = rb_class_new_instance(2, pair_LTR, cSurroGate_Pair); // SurroGate::Pair.new(right, left)
-
-  VALUE pairing = rb_iv_get(self, "@pairing");
+  int *selector;
+  VALUE scoreboard = rb_iv_get(self, "@scoreboard");
 
+  // Check the arguments for the correct type
   Check_Type(left, T_FILE);
   Check_Type(right, T_FILE);
 
-  // raise ArgumentError if @pairing.detect(&pairing_compare)
-  if (rb_block_call(pairing, rb_intern("detect"), 0, NULL, pairing_compare, rb_ary_new_from_values(2, pair_LTR)) != Qnil) {
+  // raise ArgumentError if a socket is already registered
+  if (rb_funcall(scoreboard, rb_intern("include?"), 1, left) == Qtrue || rb_funcall(scoreboard, rb_intern("include?"), 1, left) == Qtrue) {
```

Comment:
I'm having trouble seeing the difference between the first and second bit.
Should it be `left` and `right` or something?

ASIDE: do you need to call `rb_intern("include?")` twice? Is it very cheap?